### PR TITLE
Fix transpile() for control flow operations with a Target/BackendV2

### LINKED ISSUE #8824: Add support for variable width operations to Target
### What should we add?

Right now the target class doesn't have proper support for modelling instructions that have variable width. If you have a gate that can use more than a single number of qubits and the backend supports all variants of that instruction (or at least more than one). We don't have a way to express that. The target currently requires a 1:1 mapping of operation name to a `Operation` instance and that instance is used for type checking and parameter checking at different points in the transpiler. However this requirement won't work for variable width instructions because a single instance of the instructions can only have a single number of qubits.

To be able to accurately model these instructions being supported on a target backend we need to come up with an interface to model this in the target, which might mean introducing a `TargetV2`.

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit fixes an issue when compiling with circuits that have control flow operations and are running transpile() with a BackendV2 instance or a custom Target. In these cases the transpile() operation would fail to run because the Target didn't have a provision to recognize a global variable width operation as part of the target. Previously all operations in the target needed to have an instance of an Operation so that the parameters and number of qubits could be verified. Each of these operation instances would be assigned a unique name. But, for control flow operations they're defined over a variable number of qubits and all have the same name (this is a similar problem for gates like mcx too). Not being able to fully represent the control flow operations in a target was preventing running transpile() on a circuit with control flow. This commit fixes this by adding support to the target to represent globally defined operations by passing the class into Target.add_instruction instead of an instance of that class. When the class is received the Target class will treat that operation name and class as always being valid for the target for all qubits and parameters. This can then be used to represent control flow operations in the target and run transpile() with control flow operations.

### Details and comments

Fixes #8824

TODO:

- [x] Fix transpile() test (may be a bug in test code or a bug in a transpiler pass with control flow), depends on #8880 and #8879
- [x] Add tests for coupling map generation in target
- [x] Add release note

### CODE REVIEW

**Comment by mtreinish** [comment]
I believe the test failures here are being caused by a bug in `transpile()` with control flow. I'm seeing a control flow block that has a cx between two qubits that don't have connectivity in the target. Unless, I'm getting the mapping messed up in my head while debugging, I'll keep digging into this and see if I can figure out what's going on, but I'm not sure we have any full path tests with control flow that check the output circuit is valid for the backend (which this PR does) so there could have been a bug that slipped through.

**Comment by jakelishman** [comment]
The bug looks like it must be in `StochasticSwap` somewhere - I threw a callback into the test that runs `CheckMap` on the dag after every stage, and `is_swap_mapped` never gets set to `True`.

File: qiskit/transpiler/target.py [hunk 1]
```diff
@@ -319,19 +326,33 @@ def add_instruction(self, instruction, properties=None, name=None):
...
-            self._qarg_gate_map[qarg].add(instruction_name)
+        if is_class:
+            qargs_val = {None: None}
+        else:
+            qargs_val = {}
+            for qarg in properties:
+                if qarg is not None:
+                    self.num_qubits = max(self.num_qubits, max(qarg) + 1)
+                qargs_val[qarg] = properties[qarg]
+                self._qarg_gate_map[qarg].add(instruction_name)
```
Comment (jakelishman) [inline qiskit/transpiler/target.py:355]: Should we handle the case that a user passes a `properties` argument with a class?  I feel like `{None: None}` is the only thing that makes sense for control-flow, but I could potentially imagine `mcx` wanting to specify the allowed combinations of qubits directly _maybe_?

Happy just to make it a warning/error for this release, though, and think more about the interface later.

File: qiskit/transpiler/target.py [hunk 2]
```diff
@@ -507,14 +530,17 @@ def operations_for_qargs(self, qargs):
...
 
         Raises:
             KeyError: If qargs is not in target
         """
         if qargs not in self._qarg_gate_map:
             raise KeyError(f"{qargs} not in target.")
-        return [self._gate_name_map[x] for x in self._qarg_gate_map[qargs]]
+        res = [self._gate_name_map[x] for x in self._qarg_gate_map[qargs]]
+        res += [self._gate_name_map[x] for x in self._qarg_gate_map[None]]
+        return res
```
Comment (jakelishman) [inline qiskit/transpiler/target.py:543]: The `self._qarg_gate_map[None]` call here has the side-effect of adding `None` to the `defaultdict`, so `None in self._qarg_gate_map` will be `True` after this call, even if `None` wasn't originally there.  That _probably_ doesn't cause any problems, but it might be safer to avoid it.

File: qiskit/transpiler/target.py [hunk 3]
```diff
@@ -627,6 +667,25 @@ def check_obj_params(parameters, obj):
         if operation_name in self._gate_map:
             if parameters is not None:
                 obj = self._gate_name_map[operation_name]
+                if inspect.isclass(obj):
+                    warnings.warn(
+                        "A parameter was specified for an operation that is defined as a "
+                        "globally supported class there is no available validation (including "
+                        "whether the specified operation supports parameters), the returned "
+                        "value will not factor in the argument `parameters`.",
```
Comment (jakelishman) [inline qiskit/transpiler/target.py:675]: This sentence reads weirdly to me - I think it might be missing some words / a sentence break in the middle?

File: qiskit/transpiler/target.py [hunk 4]
```diff
@@ -661,10 +732,20 @@ def operations(self):
...
+        ``(class, None)`` where class is the actual operation class that
+        is globally defined.
+        """
+        res = []
+        for op in self._gate_map:
+            for qarg in self._gate_map[op]:
+                if not inspect.isclass(self._gate_name_map[op]):
+                    res.append((self._gate_name_map[op], qarg))
+                else:
+                    res.append((self._gate_name_map[op], None))
```
Comment (jakelishman) [inline qiskit/transpiler/target.py:747]: Won't the `qarg` just automatically be `None` in the lookup here?  I'm not certain we need the special handling?

File: qiskit/transpiler/target.py [hunk 5]
```diff
@@ -711,6 +792,8 @@ def _build_coupling_graph(self):
         self._coupling_graph.add_nodes_from([{} for _ in range(self.num_qubits)])
         for gate, qarg_map in self._gate_map.items():
             for qarg, properties in qarg_map.items():
+                if qarg is None:
+                    continue
```
Comment (jakelishman) [inline qiskit/transpiler/target.py:796]: It might be worth adding a release note about the changes in this function, since I think there is a user-facing API change here?

File: qiskit/transpiler/target.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/transpiler/target.py:355]: yeah, I think we should error in that case, my feeling was if properties were passed in they should be using an instance of the gate. But we can discuss this for 0.23.0 if we need something different, an exception is easy to roll back if we want to allow it in the future.

File: qiskit/transpiler/target.py (hunk 3 above)
Comment (mtreinish) [reply qiskit/transpiler/target.py:675]: I started to reword this to clean it up, but I'm actually thinking we should just drop this warning, something I've learned the hardware is warning at what is nominally an internal compiler data structure tends to not be actionably by the majority of users. For someone like me who is working on new compiler techniques and testing the compiler the target API is user facing, but for most people calling `transpile()` they never see this as it's internal between the backend and the compiler. So if a backend defines a variable width gate that has parameters this could trigger through no direct action from the user.

File: qiskit/transpiler/target.py (hunk 4 above)
Comment (mtreinish) [reply qiskit/transpiler/target.py:747]: You're right it will be, this is leftover from an earlier draft where I was thinking the properties payload for the target would be `None` instead of `{None: None}`. But I can revert this change and just use the list comprehension from before without issue.

File: qiskit/transpiler/target.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/transpiler/target.py:359]: Done in: https://github.com/Qiskit/qiskit-terra/pull/8852/commits/6ce7fe6375d34bf879ed9aac7bcdae2d3102179d

File: qiskit/transpiler/target.py (hunk 3 above)
Comment (mtreinish) [reply qiskit/transpiler/target.py]: Converted this to a comment (and reworded) in: https://github.com/Qiskit/qiskit-terra/pull/8852/commits/07ed34fb046ec7fb798406405c55aaffaf476535

File: qiskit/transpiler/target.py (hunk 2 above)
Comment (mtreinish) [reply qiskit/transpiler/target.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/8852/commits/e1b72bd35cd47cfdb55b562eca2e98e277240a1b

File: qiskit/transpiler/target.py (hunk 4 above)
Comment (mtreinish) [reply qiskit/transpiler/target.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/8852/commits/17ee512e87676059faee21a21e62c6afbec35201

File: qiskit/transpiler/target.py (hunk 5 above)
Comment (mtreinish) [reply qiskit/transpiler/target.py:794]: Added in: https://github.com/Qiskit/qiskit-terra/pull/8852/commits/f06d1b89ec38c0762df6b9352d1e21c3f1a2fad4

**Comment by mtreinish** [comment]
> It _technically_ breaks the public API for consumers of the `Target` class (since the output type signature of `Target.instructions` and a couple of others change), but I think that might be a necessary evil.

I think for `Target.instructions` the only real difference is that it will return a class instead of an instance for the new support for classes (the `None` return was still possible before for global defined ideal instances) I'm not sure that's really breaking unless someone is opting into the new feature usage. The coupling map change is more breaking imo, but the old behavior was incorrect though.
